feat(transport): add bandwidth tracking (send only) by space (#157)

anchalshivank opened the pull request:
Implemented tracking of bytes sent per space in the transport layer by wrapping implementations of the Transport trait. This enables basic observability of outbound traffic for each space.

Currently, only sent data is tracked; receiving metrics can be added in a follow-up.

Related to: #157

anchalshivank committed:
feat(bandwidth): track bandwidth usage by space and module (#157)

Introduced BandwidthStats and BandwidthTracker to record bytes sent and received per space and optionally per module. This lays the groundwork for future bandwidth limiting capabilities. Currently, the tracking is focused on outgoing and incoming traffic and supports querying stats via the tracker interface.

This addresses the reporting aspect of bandwidth limiting discussed in #157.

feat(bandwidth): track bandwidth usage by space and module (#157)

Introduced BandwidthStats and BandwidthTracker to record bytes sent and received per space and optionally per module. This lays the groundwork for future bandwidth limiting capabilities. Currently, the tracking is focused on outgoing and incoming traffic and supports querying stats via the tracker interface.

This addresses the reporting aspect of bandwidth limiting discussed in #157.

ThetaSinner commented on crates/api/src/transport.rs:
```diff
@@ -222,6 +231,8 @@ pub trait Transport: 'static + Send + Sync + std::fmt::Debug {
 
     /// Dump network stats.
     fn dump_network_stats(&self) -> BoxFut<'_, K2Result<TransportStats>>;
+    /// Get Bandwidth Tracker
```
```suggestion

    /// Get Bandwidth Tracker
```

ThetaSinner replied:
Just adding a newline

ThetaSinner commented on crates/api/src/transport.rs:
```diff
@@ -511,3 +547,172 @@ pub struct TransportConnectionStats {
+    /// # Arguments
+    ///
+    /// * `space` - The unique identifier for the space.
+    /// * `module` - Optional module identifier.
```
Maybe mention that it's required when tracking module messages but allowed to be absent when sending messages at the space level, not via a module?

Not sure how to write that with fewer works :)

anchalshivank replied:
Fixed

ThetaSinner commented on crates/core/src/factories/core_space.rs:
```diff
@@ -284,6 +284,9 @@ impl CoreSpace {
             }
         }
     }
+    pub fn get_bandwidth_tracker(&self) -> Arc<BandwidthTracker> {
```
```suggestion

    pub fn get_bandwidth_tracker(&self) -> Arc<BandwidthTracker> {
```

ThetaSinner replied:
Another newline

ThetaSinner reviewed (commented):
This looks great! I really appreciate the quality documentation on the new types. We've made an effort to try to keep up with documentation on this project and it's great if we can keep that going :)

A couple of really minor comments. It'd also be good to take the `.idea` files back out of Git please. I'm a RustRover user myself, but we try to keep IDE files out of Git because we all use different tools on this project :)

anchalshivank committed:
Update crates/api/src/transport.rs

Co-authored-by: ThetaSinner <ThetaSinner@users.noreply.github.com>

anchalshivank committed:
Update crates/core/src/factories/core_space.rs

Co-authored-by: ThetaSinner <ThetaSinner@users.noreply.github.com>

anchalshivank committed:
Address review feedback for bandwidth tracking

Resolved comments on the BandwidthTracker implementation.

anchalshivank committed:
Fix clippy errors

jost-s commented on crates/api/src/transport.rs:
```diff
@@ -189,7 +197,6 @@ pub trait Transport: 'static + Send + Sync + std::fmt::Debug {
         module: String,
         handler: DynTxModuleHandler,
     );
-
```
Please keep this empty line between functions.

jost-s commented on crates/api/src/transport.rs:
```diff
@@ -511,3 +546,178 @@ pub struct TransportConnectionStats {
+        module: Option<String>,
+        bytes: u64,
+    ) {
+        if let Ok(mut space_stats) = self.space_stats.write() {
+            space_stats
+                .entry(space.clone())
+                .or_insert_with(BandwidthStats::new)
+                .add_sent(bytes);
+        }
```
I suggest to unwrap the write lock:
```suggestion
        self.space_stats.write()
            .unwrap()
            .entry(space.clone())
            .or_insert_with(BandwidthStats::new)
            .add_sent(bytes);
```
That's what we've done throughout this repository with sync Mutexes and RwLocks and what the [examples](https://doc.rust-lang.org/std/sync/struct.RwLock.html#examples) do too.

jost-s commented on crates/api/src/transport.rs:
```diff
@@ -511,3 +546,178 @@ pub struct TransportConnectionStats {
+        module: Option<String>,
+        bytes: u64,
+    ) {
+        if let Ok(mut space_stats) = self.space_stats.write() {
```
Same here

jost-s reviewed (changes requested):
The code reads very clear. Nevertheless please add a test for this feature. If you want guidance how to do that, give us a shout. There are unit tests for most features in this repo, so you can have a look at them for inspiration.

anchalshivank committed:
Update crates/api/src/transport.rs

Co-authored-by: Jost <28270981+jost-s@users.noreply.github.com>

cdunster commented on .idea/workspace.xml:
Should this file be committed? Is there something non-standard in this config that might help others who are using IntelliJ?

cdunster commented on crates/api/src/transport.rs:
```diff
@@ -222,6 +229,9 @@ pub trait Transport: 'static + Send + Sync + std::fmt::Debug {
     /// Dump network stats.
     fn dump_network_stats(&self) -> BoxFut<'_, K2Result<TransportStats>>;
+
+    /// Get Bandwidth Tracker
```
Is it possible to elaborate a bit on this comment? Maybe something like "Get a reference to the struct that holds the stats of bandwidth usage of..."

anchalshivank replied:
Yups I will elaborate

cdunster commented on crates/api/src/transport.rs:
```diff
@@ -255,9 +266,15 @@ impl DefaultTransport {
         out
     }
+
+    /// Get Bandwidth tracker
```
Same here.

cdunster commented on crates/api/src/transport.rs:
```diff
@@ -321,14 +338,21 @@ impl Transport for DefaultTransport {
             })
             .encode()?;
-            self.imp.send(peer, enc).await
+            match self.imp.send(peer, enc).await {
+                Ok(()) => {
+                    self.tracker.track_sent(space, None, len);
+                    K2Result::Ok(())
+                }
+                Err(err) => K2Result::Err(err),
+            }
```
I think that you could use the `?` operator here:

```suggestion
            self.imp.send(peer, enc).await?;
            self.tracker.track_sent(space, None, len);
            K2Result::Ok(())
```

Or even, a more functional approach using [inspect](https://doc.rust-lang.org/std/result/enum.Result.html#method.inspect):

```suggestion
            self.imp
                .send(peer, enc)
                .await
                .inspect(|_| self.tracker.track_sent(space, None, len))
```

This is just a preference/opinion though, so it's up to you :slightly_smiling_face:

cdunster commented on crates/api/src/transport.rs:
```diff
@@ -340,20 +364,31 @@ impl Transport for DefaultTransport {
             })
             .encode()?;
-            self.imp.send(peer, enc).await
+            match self.imp.send(peer, enc).await {
+                Ok(()) => {
+                    self.tracker.track_sent(space, Some(module), len);
+                    K2Result::Ok(())
+                }
+                Err(err) => K2Result::Err(err),
+            }
```
Same here as above

cdunster reviewed (commented):
Looking good. Thanks!

I particularly like the detailed and clear comments. I added a couple of suggestions.

anchalshivank committed:
test: add test case and reformat code using cargo fmt

anchalshivank commented:
I haven't added tests for this module because the behavior is often non-deterministic. For example, if I wait for 5 seconds, the node may gossip during that time, leading to inconsistent test results. See the relevant code here: https://github.com/anchalshivank/kitsune2/blob/bc941a0f3f15b42c478fa052ca11cf4484c17b2e/crates/kitsune2/src/lib.rs#L263. Suggestions for handling this kind of testing would be appreciated.

anchalshivank committed:
Fix the failing test case due to timeout issue

jost-s commented:
> I haven't added tests for this module because the behavior is often non-deterministic. For example, if I wait for 5 seconds, the node may gossip during that time, leading to inconsistent test results. See the relevant code here: https://github.com/anchalshivank/kitsune2/blob/bc941a0f3f15b42c478fa052ca11cf4484c17b2e/crates/kitsune2/src/lib.rs#L263. Suggestions for handling this kind of testing would be appreciated.

Sure I'll make suggestions. This kind of thing is inherent to this system, where you can't guarantee that something has happened. Our go to tool for this circumstance is to set a timeout and inside of the timeout repeatedly query what should be asserted until it succeeds. If time runs out, the timeout must be longer, until it is stable on CI. Or something can be improved in the test logic. Or it can also happen that a field must be exposed just for testing purposes.

There's a macro for this kind of repeated check https://github.com/holochain/kitsune2/blob/main/crates/test_utils/src/lib.rs#L42

Here's an example of this macro in use https://github.com/holochain/kitsune2/blob/main/crates/core/src/factories/core_fetch/test/incoming_request_queue.rs#L134

Instead of the sleeps (e. g. https://github.com/anchalshivank/kitsune2/blob/bc941a0f3f15b42c478fa052ca11cf4484c17b2e/crates/kitsune2/src/lib.rs#L316), use the `iter_check` macro, and that should make the test predicable.

ThetaSinner commented:
I like Jost's suggestion for this, and to prove it's working, it's not necessary to assert a specific value I think. An `iter_check!` that asserts "sent > 500" and "received > 500" does demonstrate that the value is increasing as expected when nodes talk to each other.

I've chosen the value 500 at random there, not necessarily useful. It's just about something that is "> 0" really.

anchalshivank committed:
refactor: improve BandwidthTracker and add module test cases

- Updated `get_space_stats` to accept `Option<SpaceId>`:
  * Returns per-space stats if `Some(SpaceId)`
  * Returns aggregated stats if `None`
- Added module-level tests:
  * `track_sent_bytes_per_module_two_peer`
  * `track_sent_bytes_per_module_no_peer`

anchalshivank commented:
Made minor changes to some of the previous test cases that were failing due to timeout issues. The application code remains unchanged. I investigated the failures but couldn't identify the exact cause, so I increased the timeout duration in the tests as a workaround.

anchalshivank committed:
refactor(gossip): Add attributes to K2GossipConfig

- Merged fields from upstream Kistune p2p source (f65c4ec) and existing attributes
- Updated `Default` implementation accordingly

anchalshivank committed:
refactor(gossip): Revert K2GossipConfig attributes

- Add fields from upstream Kistune p2p source (f65c4ec) to `BandwidthConfig`

anchalshivank commented:
@ThetaSinner  @jost-s  @cdunster  I’ve added a code snippet for the Bandwidth module. I believe the way I’ve integrated it into the existing code is appropriate. Once everyone agrees on this approach, I’ll proceed with implementing the logic for configuring bandwidth.

Now that the transport layer includes both the tracker and config parameters, monitoring and control can be handled more effectively at this level.

Initially, I considered integrating the bandwidth parameters into the k2GossipModule, but I decided it would be cleaner to create a separate struct instead, without modifying the existing code.

anchalshivank commented on crates/gossip/src/harness.rs:
```diff
@@ -118,7 +118,7 @@ impl K2GossipFunctionalTestHarness {
     /// Wait for the given ops to be in our op store.
     pub async fn wait_for_ops(&self, op_ids: Vec<OpId>) -> Vec<MemoryOp> {
-        tokio::time::timeout(Duration::from_millis(500), {
+        tokio::time::timeout(Duration::from_millis(5000), {
```
The test cases for the 500 status code were failing, so I increased the timeout. However, I'm unable to debug the root cause, as I haven't made any changes to the relevant part of the code.

anchalshivank commented on crates/kitsune2/src/lib.rs:
```diff
@@ -182,7 +181,8 @@ mod test {
     #[tokio::test]
     async fn two_node_gossip() {
-        enable_tracing();
+        //If I comment it test passes else fails
+        // enable_tracing();
```
yups

anchalshivank committed:
Update crates/kitsune2/src/lib.rs

Co-authored-by: Callum Dunster <cdunster@users.noreply.github.com>

anchalshivank committed:
Update crates/transport_tx5/src/lib.rs

Co-authored-by: Callum Dunster <cdunster@users.noreply.github.com>

anchalshivank committed:
Update crates/kitsune2_showcase/src/app.rs

Co-authored-by: Callum Dunster <cdunster@users.noreply.github.com>

anchalshivank committed:
Update crates/test_utils/src/space.rs

Co-authored-by: Callum Dunster <cdunster@users.noreply.github.com>

anchalshivank committed:
Update crates/core/src/factories/mem_transport.rs

Co-authored-by: Callum Dunster <cdunster@users.noreply.github.com>

anchalshivank commented on crates/api/src/transport.rs:
```diff
@@ -101,7 +106,8 @@ impl TxImpHnd {
         } = data;
-
+        let len = data.len() as u64;
+        println!("Bandwidth config {:?}", self.config);
```
I added them for debugging purpose. As this is not the final code

jost-s commented:
Is this ready for review again, @anchalshivank?

anchalshivank commented:
> Is this ready for review again, @anchalshivank?

Nope, I will make some changes by the end of the day.

anchalshivank commented:
> @anchalshivank Are you willing to finish this PR?

Yes I will finish it by tomorrow, I have been very busy these days

jost-s commented:
@anchalshivank I see you haven't finished this yet. This is just an observation, you don't need to explain yourself or make a promise. I'm sure there is always enough to do for you. =)

Are you okay with us completing this PR?

anchalshivank commented:
I need more time . If anyone else wants to complete this pr,  you can assign it to them.

ThetaSinner commented:
No activity in a long time, closing
